feat(parser)!: use `BindingIdentifier` for `namespace` declaration names

DonIsaac opened the pull request:
Use `BindingIdentifier` instead of `IdentifierName` so that AST visitors can access the bound symbol id for the namespace's name. This makes namespace consistent with other named declarations, such as `Class`, `Function`, and `TSInterfaceDeclaration`.

We should consider modifying `TSModuleDeclarationName::StringLiteral(...)` to also store a `symbol_id`, since one gets bound in semantic for it as well.

DonIsaac commented:
* **#6004** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6004?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6003** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6003?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented:
Let me check the spec.

Boshen commented:
This is incorrect, it's not a binding identifier. Although the "global" fix is correct in the next PR.

DonIsaac commented:
How should we make a namespace's symbol is available from the AST?

Boshen commented:
> How should we make a namespace's symbol is available from the AST?

Can you elaborate?

DonIsaac commented:
> > How should we make a namespace's symbol is available from the AST?
> 
> Can you elaborate?

`BindingIdentifier` has a `symbol_id` property, but `IdentifierName` does not. In this snippet, a symbol `Foo` is bound and put into `SymbolTable`
```ts
namespace Foo {}
```
but since `TSModuleDeclarationName` is an `IdentifierReference`, `Foo`'s symbol id is not available in the AST.

Boshen commented:
> > > How should we make a namespace's symbol is available from the AST?
> > 
> > 
> > Can you elaborate?
> 
> `BindingIdentifier` has a `symbol_id` property, but `IdentifierName` does not. In this snippet, a symbol `Foo` is bound and put into `SymbolTable`
> 
> ```ts
> namespace Foo {}
> ```
> 
> but since `TSModuleDeclarationName` is an `IdentifierReference`, `Foo`'s symbol id is not available in the AST.

Let me double check the outdated tsc reference...

overlookmotel commented:
In my view, regardless of TS's spec, in our scheme, anything which has a record in the `SymbolTable` *must* have a `symbol_id` field. Without that field, our transform checker is not able to check the correctness of `SymbolTable`. I was planning to open a PR doing exactly this when I come to fix the scopes/symbols manipulation in TS transform.

I suppose we *could* add `symbol_id` field to `TSModuleDeclaration` instead, but in my view it's much simpler to make the `id` a `BindingIdentifier`, as this PR does.

So am re-opening this and, if you're willing @Boshen, would like to merge it.

Boshen commented:
I need proof and reference from tsc to confirm this is a BindingIdentifier.

DonIsaac commented:
@Boshen from section `10.1 Module Declarations` in the TypeScript v1.4 spec:

```
ModuleDeclaration:
  module IdentifierPath { ModuleBody }
IdentifierPath:
  Identifier
  IdentifierPath . Identifier
```

The TypeScript spec uses `Identifier` for all other declarations. For example, here is the grammar for `ClassDeclaration` in section `8.1 Class Declarations`:

```
ClassDeclaration:
  class Identifier TypeParametersopt ClassHeritage { ClassBody }
```

Our `Class` node uses a `BindingIdentifier`, meaning `Identifier` in TypeScript is the same as `BindingIdentifier` in oxc.

DonIsaac committed:
feat(parser)!: use `BindingIdentifier` for `namespace` declaration names (#6003)

Use `BindingIdentifier` instead of `IdentifierName` so that AST visitors can access the bound symbol id for the namespace's name. This makes namespace consistent with other named declarations, such as `Class`, `Function`, and `TSInterfaceDeclaration`.

We should consider modifying `TSModuleDeclarationName::StringLiteral(...)` to also store a `symbol_id`, since one gets bound in semantic for it as well.
